landlab/components/groundwater/dupuit_percolator.py

```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+"""
+GroundwaterDupuitPercolator Component
+
+@author: G Tucker
+"""
+
+import numpy as np
+from landlab import Component
+from landlab.utils import return_array_at_node, return_array_at_link
+from landlab.grid.mappers import map_mean_of_link_nodes_to_link
+
+
+ACTIVE_LINK = 0
+
+
+class GroundwaterDupuitPercolator(Component):
+    """
+    Simulate groundwater flow in a shallow unconfined aquifer.
+
+    The GroundwaterDupuitPercolator uses the Dupuit approximation that the
+    hydraulic gradient is equal to the slope of the water table.
+
+    Parameters
+    ----------
+    grid: ModelGrid
+            Landlab ModelGrid object
+    hydraulic_conductivity: float, field name, or array of float
+            saturated hydraulic conductivity, m/s
+            Default = 0.01 m/s
+    recharge_rate: float, field name, or array of float
+            Rate of recharge, m/s
+            Default = 1.0e-8 m/s
+
+    Examples
+    --------
+    >>> from landlab import RasterModelGrid
+    >>> mg = RasterModelGrid((3, 3))
+    >>> z = mg.add_zeros('node', 'topographic__elevation')
+    >>> gdp = GroundwaterDupuitPercolator(mg)
+
+    Notes
+    -----
+    Groundwater discharge per unit length, q, is calculated as:
+
+        q = - K H dw/dx,
+
+    where K is hydraulic conductivity, H is aquifer thickness, w is water table
+    height, and x is horizontal distance.
+
+    An explicit forward-in-time finite-volume method is used to implement a
+    numerical solution. Flow discharge between neighboring nodes is calculated
+    using the average depth at the nodes.
+
+    Note that the current version does NOT handle surface seepage.
+    """
+
+    _name = "GroundwaterDupuitPercolator"
+
+    _input_var_names = set(("topographic__elevation",
+                            "aquifer_base__elevation"))
+
```

Comment:
I think that our current conception of an input is that it must already exist before the model is initialized. Thus, I think your handling of these two inputs in the init doesn't make sense. 

They should exist (perhaps it must be asserted that the aquifer elevation is always below or equal to the topography), and if they don't a Field error gets raised.

Reply:
There's a larger issue: what do we call fields that can be created a priori, but if not the component provides a default? Or creates via create_grid? I'm going to leave this as is for now until we come up with a general resolution/recommendation.